driver/src/main/java/org/neo4j/driver/internal/net/pooling/BlockingPooledConnectionQueue.java

```diff
@@ -0,0 +1,121 @@
+/**
+ * Copyright (c) 2002-2016 "Neo Technology,"
+ * Network Engine for Objects in Lund AB [http://neotechnology.com]
+ *
+ * This file is part of Neo4j.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.neo4j.driver.internal.net.pooling;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Set;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.LinkedBlockingQueue;
+
+import org.neo4j.driver.internal.util.Supplier;
+
+/**
+ * A blocking queue that also keeps track of connections that are acquired in order
+ * to facilitate termination of all connections.
+ */
+public class BlockingPooledConnectionQueue
+{
+    /** The backing queue, keeps track of connections currently in queue */
+    private final BlockingQueue<PooledConnection> queue;
+
+    /** Keeps track of acquired connections */
+    private final Set<PooledConnection> acquiredConnections =
+            Collections.newSetFromMap(new ConcurrentHashMap<PooledConnection, Boolean>());
+
+    public BlockingPooledConnectionQueue( int capacity )
+    {
+        this.queue = new LinkedBlockingQueue<>( capacity );
+    }
+
+    /**
+     * Offer a connections back to the queue
+     *
+     * @param pooledConnection the connection to put back to the queue
+     * @return <code>true</code> if connections was accepted otherwise <code>false</code>
+     */
+    public boolean offer( PooledConnection pooledConnection )
+    {
+        boolean offer = queue.offer( pooledConnection );
+        if ( offer )
+        {
+            acquiredConnections.remove( pooledConnection );
+        }
+        return offer;
+    }
+
+    public List<PooledConnection> toList()
+    {
+        return new ArrayList<>( queue );
+    }
+
+    public boolean isEmpty()
+    {
+        return queue.isEmpty();
+    }
+
+    public int size()
+    {
+        return queue.size();
+    }
+
+    public boolean contains( PooledConnection pooledConnection )
+    {
+        return queue.contains( pooledConnection );
+    }
+
+    /**
+     * Terminates all connections, both those that are currently in the queue as well
+     * as those that have been acquired.
+     */
+    public void terminate()
```

Comment:
You probably want a atomic var here at the very beginning of the method, to flag it that we've started termination. If any connection created after this, close it right now. If any connection want to _add_ to the queue or map, drop them directly.